homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,294 @@
+    _available = True
+
+    def __init__(self, device_id, device_port, initial_event=None,
+                 name=None):
+        """Initialize the device."""
+        # HLK-SW16 specific attributes for every component type
+        self._initial_event = initial_event
+        self._device_id = device_id
+        self._device_port = str(device_port)
+        self._is_on = None
+        if name:
+            self._name = name
+        else:
+            self._name = device_id
+
+    @callback
+    def handle_event_callback(self, event):
+        """Propagate changes through ha."""
+        self.async_schedule_update_ha_state()
+
+    def _handle_event(self, event):
+        """Platform specific event handler."""
+        raise NotImplementedError()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return a name for the device."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._is_on
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @callback
+    def _availability_callback(self, availability):
+        """Update availability state."""
+        self._available = availability
+        self.async_schedule_update_ha_state()
+
+    async def async_added_to_hass(self):
+        """Register update callback."""
+        async_dispatcher_connect(self.hass, SIGNAL_AVAILABILITY,
+                                 self._availability_callback)
+
+        # Process the initial event now that the entity is created
+        if self._initial_event:
+            self.handle_event_callback(self._initial_event)
+
+
+class SW16Command(SW16Device):
+    """Singleton class to make HLK-SW16 command interface available to entities.
+
+    This class is to be inherited by every Entity class that is actionable
+    (switches/lights). It exposes the HLK-S16 command interface for these
+    entities.
+
+    The HLK-SW16 interface is managed as a class level and set during setup
+    (and reset on reconnect).
+    """
+
+    _protocol = None
+
+    @classmethod
+    def set_hlk_sw16_protocol(cls, protocol):
+        """Set the HLK-S16 asyncio protocol as a class variable."""
+        cls._protocol = protocol
+
+    @classmethod
+    def is_connected(cls):
+        """Return connection status."""
+        return bool(cls._protocol)
+
+    async def _async_handle_command(self, command, *args):
+        """Do bookkeeping for command, send it to HLK-SW16 and update state."""
+        if not self.is_connected():
+            raise HomeAssistantError('Cannot send command, not connected!')
+
+        if command == 'turn_on':
+            state = await self._protocol.turn_on(codecs.decode(
+                self._device_port.rjust(2, '0'), 'hex'))
+            _LOGGER.debug("new state:")
+            _LOGGER.debug(state)
```

Comment:
Just log once per occasion.

Reply:
cleaned this up so it also logs the relay